File: src/app/core/services/categories/category.service.ts
import { HttpClient, HttpParams, HttpHeaders } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { environment } from 'src/environments/environment';
import { BaseResponse } from '../../models/response/base-response';
import { CategoryResponse } from '../../models/response/Category/CategoryResponse';
import { PageResponse } from '../../models/response/page-response';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';
import { CategoryRequest } from '../../models/request/Category/CategoryRequest';

@Injectable({
  providedIn: 'root'
})
export class CategoryService {

  private apiUrl = `${environment.apiUrl}/categories`;

  constructor(private http: HttpClient) { }

  getAllCategories(pageNumber: number, pageSize: number): Observable<BaseResponse<PageResponse<CategoryResponse>>> {
    const params = new HttpParams()
      .set('pageNumber', pageNumber)
      .set('pageSize', pageSize);
    
    return this.http.get<BaseResponse<PageResponse<CategoryResponse>>>(this.apiUrl, { params });
  }

  getCategoryDetails(id: number): Observable<BaseResponse<CategoryResponse>> {
    return this.http.get<BaseResponse<CategoryResponse>>(`${this.apiUrl}/details/${id}`);
  }

  adminCreateCategory(request: CategoryRequest): Observable<BaseResponse<any>> {
    return this.http.post<BaseResponse<any>>(`${this.apiUrl}/add`, request);
  } 

  updateCategory(id: number, request: CategoryRequest): Observable<BaseResponse<any>> {
    return this.http.patch<BaseResponse<any>>(`${this.apiUrl}/${id}`, request);
  }

  deleteCategory(id: number): Observable<BaseResponse<any>> {
    return this.http.delete<BaseResponse<any>>(`${this.apiUrl}/${id}`);
  }

  countCategories(): Observable<number> {
    return this.http.get<BaseResponse<number>>(`${this.apiUrl}/count`)
      .pipe(map(res => res.data));
  }

  /**
   * Export categories report as a Blob (e.g., PDF)
   */
  exportCategoryReport(): Observable<Blob> {
    return this.http.get(`${environment.apiUrl}/reports/categories`, {
      responseType: 'blob'
    });
  }
}

File: demo/src/main/java/com/training/demo/service/CategoryService.java
package com.training.demo.service;

import com.training.demo.dto.request.Category.CategoryRequest;
import com.training.demo.dto.response.Category.CategoryResponse;
import com.training.demo.dto.response.System.PageResponse;

public interface CategoryService {

    /**
     * Add a new category
     * @param categoryRequest the category request object
     * @return the created category response object
     */
    CategoryResponse addCategory(CategoryRequest categoryRequest);

    /**
     * Update an existing category
     * @param categoryRequest the category request object
     * @return the updated category response object
     */
    CategoryResponse updateCategory(Long id, CategoryRequest categoryRequest);

    /**
     * Delete a category by its ID
     * @param categoryId the ID of the category to be deleted
     */
    void deleteCategory(Long categoryId);

    /**
     * Get all categories with pagination
     * @param page the page number
     * @param size the size of the page
     * @return a paginated response of category responses
     */
    PageResponse<?> getAllCategories(int page, int size);

    /**
     * Count total categories
     * @return total number of categories
     */
    long countCategories();
}
